### DISCUSSION TITLE
How can I pass additional parameters to signIn() function?

### DISCUSSION THREAD

**Discussion by leodrez** [nextauthjs/next-auth#901] [Help]
**Your question**
How can I pass additional parameters to signIn with email function to use on callback?
Cant find passed values on the sign in callback

**What are you trying to do**
Im trying to add an additional "name" parameter to the signIn() function, along with the email and callback url. 
Currently Im using the prisma adapter and would like to add a value to the native "name" property that the schema requires you to have.

**Feedback**
https://next-auth.js.org/schemas/adapters
https://next-auth.js.org/getting-started/client#signin

`signIn("email", {
    email: email,
    callbackUrl: userCallbackUrl,
    name: 'Test Name',
})`

* [x] Could not find relevant documentation

**Comment by balazsorban44** [comment]
Hi, would https://github.com/nextauthjs/next-auth/pull/823 solve this issue?

**Reply by ahallora** [reply]
@leodrez did you figure this out? 

Say we do this:

```javascript
  signIn(
    "email",
    {
      email: email,
      callbackUrl: userCallbackUrl,
    },
    {
      name: "Test Name",
      locale: "en"
    }
  );
```

How (andwhere in the NextAuth-code) would I be able to retrieve that name? E.g. getting the locale of the user, so we can send a localized email, that would be great. 😅 Any ideas @balazsorban44 ? 🙏

**Reply by martinatwainobicom** [reply]
@ahallora I had exactly the same use case => send localized emails. 

While it would be great to have the parameters in the signIn Callback (not only for localized emails but also for knowing whether I am in a registration flow or login flow) I couldn't find them anywhere.

My (quite hacky) solution was to catch the parameters before NextAuth directly from the request. 

In [...nextauth].ts instead of 

```js
export default (req: NextApiRequest, res: NextApiResponse<any>) => NextAuth(req, res, options as any);
```

you can write:


```js
export default (req: NextApiRequest, res: NextApiResponse<any>) => {
	cachePotentialEmailLocaleMapping(req);
	return NextAuth(req, res, options as any);
}
```

The cachePotentialEmailLocaleMapping function then checks whether the request is exactly a signin with an email (from req.query) , reads the parameters you provided in the signIn function (third argument) on the client and caches them in an email=>locale map.

```js
function cachePotentialEmailLocaleMapping(req: NextApiRequest) {
	if(req.query.nextauth
		&& req.query.nextauth.length === 2
		&& req.query.nextauth[0] === 'signin'
		&& req.query.nextauth[1] === 'email'
		&& req.query.locale
		&& req.body.email
		&& EmailValidator.validate(req.body.email)) 
	{
		let locale = nextConfig.i18n.defaultLocale;
		if(req.query.locale !== '' 
			&& nextConfig.i18n.locales.includes(req.query.locale)) 
		{
			locale = req.query.locale;
		}
		EmailLocaleMap[req.body.email] = locale;
	}
}
```

I then use a custom sendVerificationRequest function to find the cached locale for a given email address and can send a localized verification email. Afterwards I delete the cached entry again.

I hope that helped.

**Reply by ahallora** [reply]
Thank you, @martinatwainobicom for your input. 🙌 😄 
I came to a very similar solution described here: https://github.com/nextauthjs/next-auth/issues/2143 🙈

**Reply by iandjx** [reply]
The solution doesnt seem to work with facebook login. the extra params i included are not inside req.query
```
signIn(
                  'facebook',
                  {
                    callbackUrl: `${courseURL}/?login=true`,
                    hello: 'asdas',
                  },
                  {
                    hello: '123',
                    // allowSignup,
                  }
                )```

**Reply by daniellin215** [reply]
@ahallora how were you able to get the `locale` into `req.query`? As far as I know, using EmailProvider I could specify `/api/auth/email?locale=abc` but this doesn't work for me, it does not trigger the default email provider next auth route handler or the callback logic

**Reply by ahallora** [reply]
I simply call the endpoint `/api/auth/signin/email?locale=da&callbackUrl=https://...`

Everything else is described in this thread: https://github.com/nextauthjs/next-auth/issues/2143#issuecomment-879376946 

Note that things might be a bit different from `v3` and up: https://github.com/nextauthjs/next-auth/issues/2143#issuecomment-949675729

**Reply by gorkemsevim** [reply]
> 

For facebook,

i created a new variable as a global for test,
and inside FacebookProvider obj , add profile method based on https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/providers/facebook.ts, and add my own props here like city, name whatever you want. you can get these new updated values in signIn method.

---GLOBAL OBJ------
const userAdditionInfo: {
  phoneNumber: string | null;
  city: string | null;
  name: string | null;
  surname: string | null;
} = {
  phoneNumber: null,
  city: null,
  name: null,
  surname: null,
};


----before NextAuth func----
export default async function auth(req: NextApiRequest, res: NextApiResponse) {
// update an object with req query params
  if (
    req.query.phoneNumber &&
    req.query.city &&
    req.query.name &&
    req.query.surname
  ) {
    userAdditionInfo.phoneNumber = req.query.phoneNumber as string;
    userAdditionInfo.city = req.query.city as string;
    userAdditionInfo.name = req.query.name as string;
    userAdditionInfo.surname = req.query.surname as string;
  }

return await NextAuth(req, res, {
    ...
    FacebookProviders({
     ....,
        profile(profile) {
                  return {
                    id: profile.id,
                    email: profile.email,
                    image: profile.picture.data.url,
                    phoneNumber: userAdditionInfo.phoneNumber,
                    name: userAdditionInfo.name ?? profile.name,
                    surname: userAdditionInfo.surname,
                    city: userAdditionInfo.city,
                  };
                },
        })
    }
}

**Comment by krishna-sharma-dev** [comment]
There is a temporary workaround that can be used to achieve this!

Before this handler in your file:
`signIn(provider.id, { callbackUrl: redirectUrl });`

Add a method that sets a cookie:
`setCookie(
                      'additionalAuthParams',
                      JSON.stringify({
                        appPublicKey: appPublicKey,
                      })
                    );`
like this: 
```
<Button  onClick={() => {
                    setCookie(
                      'additionalAuthParams',
                      JSON.stringify({
                        appPublicKey: appPublicKey,
                      })
                    );
                  signIn(provider.id, { callbackUrl: redirectUrl });
                }}
              >
                Sign in with {provider.name}
 </Button>
```

Now in [...nextauth].ts, we need to make change such that authOptions is just a function that returns next-auth callbacks and config like: 
```
export default async function auth(req: NextApiRequest, res: NextApiResponse) {
    return await NextAuth(req, res, {
  providers: [
   ...
  ],
  callbacks: {
    ...
    async signIn({ user, account, profile, email, credentials }) {
      let additionalAuthParams = JSON.parse(req.cookies.additionalAuthParams)
          .appPublicKey,
    },
  },
});
}```

You should be able to read the cookie and get it working

**Reply by jacekkolasa** [reply]
why is this downvoted? It's the only working solution I found so far

**Comment by pbell23** [comment]
Any update on this ?
I also need to pass custom parameters to the signin/signup using the email provider.

Here is an example of use : Allow a new user to specify its username on signup. Currently, I believe I have to do this in a separate step.

EDIT:
I just noticed that we can retrieve the parameters passed as the 3rd argument of the signIn function inside the query object of the request.

So I can just do that : 
`await signIn(
      'email',
      { email, redirect: false },
      { username: 'example-username' }
    );`

And inside the signIn callback : 
`async signIn({ user, account, profile, email, credentials }) {
        console.log(${req.query.username}) // example-username
}`

To access the request object, you need to use the nextAuth advanced initialization : [https://next-auth.js.org/configuration/initialization](https://next-auth.js.org/configuration/initialization)

**Reply by bsplosion** [reply]
This is the answer to OP's question.  Direct link to the documentation: https://next-auth.js.org/getting-started/client#additional-parameters

@leodrez want to mark this as the answer?

**Reply by trungtin** [reply]
This doesn't seem to work with Authjs (next-auth v5 beta).

**Reply by arxk** [reply]
This wont work with any old metadata, only works with things like the scope, prompt, state etc..
We need a way to safely access metadata sent from the signIn.

https://github.com/nextauthjs/next-auth/blob/3833baafdcabe0c439f020d96d79d3fe8dd1c98b/packages/core/src/lib/actions/signin/authorization-url.ts

**Reply by bsplosion** [reply]
@trungtin I'm running AuthJS/NextAuth v5 right now and leveraging this exact functionality for several parameters required during a sign in process.  The parameters do need to be encodable to the URL, but otherwise it works exactly as described.

@arxkdev this ticket is specifically covering access to params provided in the `signIn` function during a callback, which the solution above does just fine.  If you need params from/to somewhere else you might want to create a new ticket.

**Comment by Amir-Kermanshahani** [comment]
I'm not able to pass any parameters using the above solutions. any workarounds as of today?

**Reply by Amir-Kermanshahani** [reply]
I couldn't find an applicable answer for my case, but I could figure out the puzzle pieces and now it works as expected:

My current stack is: 

- "next": "14.2.3",
- "next-auth": "^4.24.7",
- "react": "^18",
- "@auth/prisma-adapter": "^2.2.0",
- "@prisma/client": "^5.14.0",

1. when calling the signIn function, the additional parameters should be sent as the third parameter, as in the [Docs](https://next-auth.js.org/getting-started/client#additional-parameters), @bsplosion thanks for mentioning that!
`await signIn(
        "email",
        {
          email: values.email,
        },
        {
          name: "Test Name"
        }
      );`

2. I'm using App router, so I implemented the Advanced initialization mentioned in the [Docs](https://next-auth.js.org/configuration/initialization#advanced-initialization), passing `NextApiRequest`, `NextApiResponse` to NextAuth options for further use and exported the main `auth` function as `GET `and `POST`.

3. I was able to get the parameters in the signIn callback function by parsing the `req.url` and using `url.searchParams.get()`

here is my complete code for `api/auth/[...nextauth]/route.ts`

```
const prisma = new PrismaClient()

async function auth(req: NextApiRequest, res: NextApiResponse) {
  return await NextAuth(req, res, {
    adapter: PrismaAdapter(prisma) as Adapter,
    providers: [...],
    pages: {
      signIn: "/signup",
    },
    callbacks: {
      async signIn({ user, account, profile, email, credentials }) {
        try {
          const url = new URL(req.url || "/signIn");
          const name= url.searchParams.get("name");
        } catch (error) {
          console.log(error);
        }
        return true;
      },
    },
  });
}

export { auth as GET, auth as POST };
```

Hope it helps, cheers!

**Reply by Waylon-Firework** [reply]
@Amir-Kermanshahani 

I can't get req and res in the signIn callbacks. Is our version different? 
This is the official example. https://github.com/nextauthjs/next-auth-example/blob/main/auth.ts


I don't know how to config the project. Do you mean we need to move the config from auth.ts file to route.ts ?

**Reply by Amir-Kermanshahani** [reply]
No need to move the config, you can keep it in a seperate auth.ts file, but NextAuth docs does it in the route.ts file.

After configuring your auth.ts file and exporting it as in the link you send, you need to import your handlers from auth.ts in the [route.ts](https://github.com/nextauthjs/next-auth-example/blob/main/app/api/auth/%5B...nextauth%5D/route.ts) file, then export the `GET` and `POST` methods to be used as API handlers for NextAuth application. 

Make sure you have initialized NextAuth in Advanced mode. (NextAuth [docs](https://next-auth.js.org/configuration/initialization#advanced-initialization))

**Reply by Mohamed-Abdulla** [reply]
@Amir-Kermanshahani 

 await signIn(
        "google",
        {
          callbackUrl: `/${redirect}`,
        },
        {
          role: "nutritionist",
        }
      );
      
async function auth(req: NextApiRequest, res: NextApiResponse) {
  return await NextAuth(req, res, {
    session: {
      strategy: "jwt",
    },

    pages: {
      signIn: "/",
    },
    providers: [
      GoogleProvider({
        clientId: getGoogleCredentials().clientId,
        clientSecret: getGoogleCredentials().clientSecret,
      }),
    ],
    callbacks: {
      async signIn({ user }) {
        const url = new URL(req.url || "/signIn");
        const role = url.searchParams.get("role");

        console.log("role", role);
        if (user.email) {
          const existingUser = await getUserByEmail(user.email);

          if (!existingUser) {
            await createUser({
              email: user.email,
              name: user.name!,
              role: role as UserRole,
              profile_picture: user.image!,
            });

            return true;
          }
        }

        return true;
      },
      
      
      the role is null for me , I couldn't able to get it. im using app router with same version as u (next, next-auth) . any help?

**Comment by scsskid** [comment]
for strava provider with nextauth5 beta, I can pass parameters like this (they will be sent as get params in url to strava when initation oauth), maybe it works similar with other providers

```js
providers: [
Strava({
	clientId: process.env.STRAVA_CLIENT_ID,
	clientSecret: process.env.STRAVA_CLIENT_SECRET,
	authorization: {
		params: { scope: 'read,activity:read' }
	}
})
],
```

**Reply by dinu7172** [reply]
can you eleborate this like how can I use this?

**Comment by Ricka7x** [comment]
is this fixed already in V5?

**Comment by dyingg** [comment]
It doesn't work as a URL parameter but you can leverage cookies and pass it as a cookie then access it from `req`.

Example:

```
document.cookie = `test=121313; domain=${
      window.location.hostname
    }; path=/; samesite=Lax`;
```

spent an awful lot of time trying to do this in a cleaner matter this use case should be better handled

**Comment by robertkibet** [comment]
For those using`auth0`, this is how to send user to respective login or signup screens with `next-auth`

```js
const isSignUp = true; // some state value to handle signup or login

// toggle nextauth
signIn(
  "auth0",
  { callbackUrl },
  { screen_hint: isSignUp ? "signup" : "login" }
)
```

```ts
signIn<RedirectableProviderType | undefined>(provider?: LiteralUnion<BuiltInProviderType> | undefined, options?: SignInOptions, authorizationParams?: SignInAuthorizationParams): Promise<SignInResponse | undefined>
```

**Comment by gideondakore** [comment]
Is it solved? And if yes, how do you do it?
